### PR TITLE
[issue #1748] Support phaseType change in invoices dry runs

### PR DESCRIPTION
**Pull request by dawidworobiec**

### CODE REVIEW

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        final Entitlement createdEntitlement = entitlementApi.getEntitlementForId(createdEntitlementId, callContext);
+        assertListenerStatus();
+
+        final LocalDate futureDate = new LocalDate(2022, 7, 13);
+        final DryRunArguments dryRunSubscriptionActionArg = new TestDryRunArguments(DryRunType.SUBSCRIPTION_ACTION,
+                                                                                    "Gun",
+                                                                                    ProductCategory.BASE,
+                                                                                    BillingPeriod.MONTHLY,
+                                                                                    null,
+                                                                                    PhaseType.EVERGREEN,
```
Comment (sbrossie) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java:69]: Could you do 2 dry-runs, one with null target phase and one with the `EVERGREEN` - that way it becomes clear the code works as expected?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    @Override
+    protected KillbillConfigSource getConfigSource(final Map<String, String> extraProperties) {
+        final Map<String, String> allExtraProperties = new HashMap<String, String>(extraProperties);
+        allExtraProperties.put("org.killbill.catalog.uri", "catalogs/testCatalogForcedPhaseChange");
+        return super.getConfigSource(null, allExtraProperties);
+    }
+
+    @Test(groups = "slow")
+    public void testDryRunWithChangePlanInMiddleOfSubscriptionWithPhaseSpecifiedOnNewCatalog() throws Exception {
```
Comment (sbrossie) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java:52]: Why `xxxOnNewCatalog` considering there is only one version?

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java [hunk 3]
```diff
@@ -30,14 +30,7 @@
...
 import org.killbill.billing.callcontext.InternalTenantContext;
-import org.killbill.billing.catalog.api.BillingActionPolicy;
-import org.killbill.billing.catalog.api.CatalogApiException;
-import org.killbill.billing.catalog.api.Plan;
-import org.killbill.billing.catalog.api.PlanChangeResult;
-import org.killbill.billing.catalog.api.PlanPhasePriceOverridesWithCallContext;
-import org.killbill.billing.catalog.api.PlanPhaseSpecifier;
-import org.killbill.billing.catalog.api.ProductCategory;
-import org.killbill.billing.catalog.api.StaticCatalog;
+import org.killbill.billing.catalog.api.*;
```
Comment (sbrossie) [inline subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java:33]: (same here): Can you add the correct imports (no *) ?

**Review by sbrossie — CHANGES_REQUESTED** [review]
The code change looks good and 👍 for adding some good `beatrix` level tests. I have a few comments for you to look at - also only looked at one test as I had a couple of remarks.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+package org.killbill.billing.beatrix.integration;
+
+import com.google.common.collect.ImmutableList;
+import org.joda.time.LocalDate;
+import org.killbill.billing.account.api.Account;
+import org.killbill.billing.api.TestApiListener.NextEvent;
+import org.killbill.billing.beatrix.util.InvoiceChecker.ExpectedInvoiceItemCheck;
+import org.killbill.billing.catalog.DefaultPlanPhasePriceOverride;
+import org.killbill.billing.catalog.api.*;
```
[the remark this answers is not in the document]
Comment (dawidworobiec) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java]: Changed

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+import org.killbill.billing.invoice.api.DryRunType;
+import org.killbill.billing.invoice.api.Invoice;
+import org.killbill.billing.invoice.api.InvoiceItemType;
+import org.killbill.billing.payment.api.PluginProperty;
+import org.killbill.billing.platform.api.KillbillConfigSource;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.math.BigDecimal;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (dawidworobiec) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java]: Changed

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java (hunk 3 above)
Comment (dawidworobiec) [reply subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java]: Changed

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java (hunk 1 above)
Comment (dawidworobiec) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java]: Hmmm, I already did 4 tests:
- at start of subscription with the phase specified
- at start of subscription without phase specified **(null target phase)**
- in middle of subscription with phase specified
- in middle of subscription without phase specified

I think it covers your use questions. Or maybe you prefer to have it inside a single unit test?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java (hunk 2 above)
Comment (dawidworobiec) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java]: Ah sorry, I left this accidentally after copying one of your tests. Removed.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    protected KillbillConfigSource getConfigSource(final Map<String, String> extraProperties) {
+        final Map<String, String> allExtraProperties = new HashMap<String, String>(extraProperties);
+        allExtraProperties.put("org.killbill.catalog.uri", "catalogs/testCatalogForcedPhaseChange");
+        return super.getConfigSource(null, allExtraProperties);
+    }
+
+    @Test(groups = "slow")
+    public void testDryRunWithChangePlanInMiddleOfSubscriptionWithPhaseSpecifiedOnNewCatalog() throws Exception {
+        final LocalDate initialDate = new LocalDate(2022, 6, 28);
+        clock.setDay(initialDate);
```
[the remark this answers is not in the document]
Comment (dawidworobiec) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestCatalogForcedPhaseChangeWithDryRun.java:60]: Ah sorry. I set the wrong effective date in XML catalog. Changed it to be in past.

**Review by sbrossie — APPROVED** [review]

**Pull request merged**